org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinCoroutineFilter.java

```diff
@@ -72,32 +73,43 @@ private void match(final MethodNode methodNode,
 			ignore.add(s);
 			ignore.add(cursor);
 
-			for (AbstractInsnNode i = methodNode.instructions
-					.getFirst(); i != null; i = i.getNext()) {
+			int suspensionPoint = 1;
+			for (AbstractInsnNode i = cursor; i != null
+					&& suspensionPoint < s.labels.size(); i = i.getNext()) {
 				cursor = i;
 				nextIsVar(Opcodes.ALOAD, "COROUTINE_SUSPENDED");
 				nextIs(Opcodes.IF_ACMPNE);
+				AbstractInsnNode continuationAfterLoadedResult = cursor;
```

Comment:
At this point we know that cursor is `JumpInsnNode`. Cast happens later, which confused me. Also `continuationAfterLoadedResult` first points to jump insn and then later gets replaced with jump target. Both can be fixed if we replace this line with:


	if (cursor == null) {
		continue;
	}
	final AbstractInsnNode continuationAfterLoadedResult = skipNonOpcodes(((JumpInsnNode) cursor).label);

Reply:
Believe me - this is exactly how first implementation looked, then I decided to make flow linear 😆

Reply:
Fixed.